File: src/main/java/me/miki/shindo/utils/translate/Translator.java
package me.miki.shindo.utils.translate;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import me.miki.shindo.utils.TimerUtils;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.StringJoiner;

public class Translator {

    public static final String AUTO_DETECT = "";
    public static final String ENGLISH = "en";
    public static final String JAPANESE = "ja";
    public static final String CHINESE_SIMPLIFIED = "zh-Hans";
    public static final String CHINESE_TRADITIONAL = "zh-Hant";
    public static final String POLISH = "pl";
    private static String authCache;
    private static TimerUtils timer;

    private static String auth() throws Exception {

        if (timer == null) {
            timer = new TimerUtils();
        }

        if (timer.delay(300 * 1000) || authCache == null) {

            URL url = new URL("https://edge.microsoft.com/translate/auth");
            HttpURLConnection con = (HttpURLConnection) url.openConnection();
            con.setRequestMethod("GET");

            BufferedReader in = new BufferedReader(new InputStreamReader(con.getInputStream()));
            String inputLine;
            StringBuilder content = new StringBuilder();

            while ((inputLine = in.readLine()) != null) {
                content.append(inputLine);
            }

            in.close();

            authCache = content.toString();

            return authCache;
        }

        return authCache;
    }

    public static String translate(String text, String from, String to) throws Exception {

        URL url = new URL("https://api.cognitive.microsofttranslator.com/translate?from=" + from + "&to=" + to + "&api-version=3.0&includeSentenceLength=true");
        HttpURLConnection con = (HttpURLConnection) url.openConnection();
        con.setRequestMethod("POST");
        con.setRequestProperty("Content-Type", "application/json");
        con.setRequestProperty("authorization", "Bearer " + auth());

        StringJoiner sj = new StringJoiner(",", "[", "]");
        sj.add("{\"Text\":\"" + text + "\"}");

        String jsonInputString = sj.toString();

        con.setDoOutput(true);
        OutputStream os = con.getOutputStream();
        byte[] input = jsonInputString.getBytes(StandardCharsets.UTF_8);
        os.write(input, 0, input.length);

        BufferedReader br = new BufferedReader(new InputStreamReader(con.getInputStream(), StandardCharsets.UTF_8));
        String responseLine;
        StringBuilder responseContent = new StringBuilder();
        while ((responseLine = br.readLine()) != null) {
            responseContent.append(responseLine.trim());
        }
        br.close();

        JsonArray jsonArray = JsonParser.parseString(responseContent.toString()).getAsJsonArray();

        StringBuilder sb = new StringBuilder();
        for (JsonElement json : jsonArray) {
            JsonObject jsonObject = json.getAsJsonObject();
            JsonArray translations = jsonObject.getAsJsonArray("translations");
            for (JsonElement trans : translations) {
                JsonObject translation = trans.getAsJsonObject();
                sb.append(translation.get("text").getAsString());
                sb.append(",");
            }
        }

        return sb.toString().replaceAll(",$", "");
    }
}

File: src/main/kotlin/me/miki/shindo/utils/translate/Translator.kt
package me.miki.shindo.utils.translate

import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import me.miki.shindo.utils.TimerUtils
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets
import java.util.StringJoiner

object Translator {

    const val AUTO_DETECT = ""
    const val ENGLISH = "en"
    const val JAPANESE = "ja"
    const val CHINESE_SIMPLIFIED = "zh-Hans"
    const val CHINESE_TRADITIONAL = "zh-Hant"
    const val POLISH = "pl"

    private var authCache: String? = null
    private var timer: TimerUtils? = null

    @Throws(Exception::class)
    private fun auth(): String {
        if (timer == null) {
            timer = TimerUtils()
        }

        if (timer!!.delay(300 * 1000) || authCache == null) {
            val url = URL("https://edge.microsoft.com/translate/auth")
            val con = url.openConnection() as HttpURLConnection
            con.requestMethod = "GET"

            val content = StringBuilder()
            BufferedReader(InputStreamReader(con.inputStream)).use { reader ->
                var line: String?
                while (reader.readLine().also { line = it } != null) {
                    content.append(line)
                }
            }

            authCache = content.toString()
            return authCache!!
        }

        return authCache!!
    }

    @JvmStatic
    @Throws(Exception::class)
    fun translate(text: String, from: String, to: String): String {
        val url = URL("https://api.cognitive.microsofttranslator.com/translate?from=$from&to=$to&api-version=3.0&includeSentenceLength=true")
        val con = url.openConnection() as HttpURLConnection
        con.requestMethod = "POST"
        con.setRequestProperty("Content-Type", "application/json")
        con.setRequestProperty("authorization", "Bearer ${auth()}")

        val sj = StringJoiner(",", "[", "]")
        sj.add("{\"Text\":\"$text\"}")
        val jsonInputString = sj.toString()

        con.doOutput = true
        con.outputStream.use { os: OutputStream ->
            val input = jsonInputString.toByteArray(StandardCharsets.UTF_8)
            os.write(input, 0, input.size)
        }

        val responseContent = StringBuilder()
        BufferedReader(InputStreamReader(con.inputStream, StandardCharsets.UTF_8)).use { br ->
            var responseLine: String?
            while (br.readLine().also { responseLine = it } != null) {
                responseContent.append(responseLine!!.trim())
            }
        }

        val jsonArray: JsonArray = JsonParser.parseString(responseContent.toString()).asJsonArray

        val sb = StringBuilder()
        for (json: JsonElement in jsonArray) {
            val jsonObject: JsonObject = json.asJsonObject
            val translations: JsonArray = jsonObject.getAsJsonArray("translations")
            for (trans: JsonElement in translations) {
                val translation = trans.asJsonObject
                sb.append(translation["text"].asString)
                sb.append(",")
            }
        }

        return if (sb.isNotEmpty()) sb.substring(0, sb.length - 1) else ""
    }
}
